pipeline/metadata/flatten_base.py

```diff
@@ -20,6 +23,87 @@
 ]
 
 
+def get_common_name(cert_name: x509.Name) -> Union[None, str]:
+  """Get the Common Name of a certificate subject or issuer.
+
+  Args:
+    cert_name: x509.Name representing a certificate subject or issuer
+
+  Returns:
+    Common Name as a string
+  """
+  try:
+    attributes = cert_name.get_attributes_for_oid(x509.oid.NameOID.COMMON_NAME)
+    if attributes:
+      return attributes[0].value
+  except x509.AttributeNotFound:
+    logging.warning('x509.AttributeNotFound: Common Name\n')
+  return None
+
+
+def get_alternative_names(cert: x509.Certificate) -> List[str]:
+  """Get the Subject Alternative Names of a certificate.
+
+  Args:
+    cert: x509.Certificate containing parsed certificate fields
+
+  Returns:
+    list of alternative names
+  """
+  try:
+    ext = cert.extensions.get_extension_for_oid(
+        x509.ExtensionOID.SUBJECT_ALTERNATIVE_NAME)
+    # Cast to x509.SubjectAlternativeName to avoid mypy error.
+    san_ext = cast(x509.SubjectAlternativeName, ext.value)
+    return san_ext.get_values_for_type(x509.DNSName)
+  except x509.extensions.ExtensionNotFound:
+    return []
+
+
+def load_cert_from_str(cert_str: str) -> x509.Certificate:
+  """Load certificate from the certificate text base64 string.
+
+  Args:
+    cert_str: base64 encoded certificate string
+
+  Returns:
+    x509.Certificate containing parsed certificate fields (e.g., Common Name)
+  """
+  begin = "-----BEGIN CERTIFICATE-----\n"
+  end = "\n-----END CERTIFICATE----- "
+  # cryptography.x509 requires the PEM headers and footers to parse the certificate.
+  cert_pem = bytes(begin + cert_str + end, encoding='utf-8')
+  return x509.load_pem_x509_certificate(cert_pem, default_backend())
+
+
+def parse_cert(cert_str: str) -> Dict[str, Any]:
+  """Parse certificate fields from base64 certificate string.
+
+  Args:
+    cert_str: base64 encoded certificate string
+
+  Returns:
+    dict containing parsed certificate fields
+  """
+  cert_fields: Dict[str, Any] = {
```

Comment:
I don't think we gain much here from having a stringly-typed dict

I would just return the fields as a tuple with the order defined in the docstring

`return (cert_common_name, cert_issuer, cert_start_date, cert_end_date, cert_alternative_name)`

and in `parse_received_headers` just read them out as

`(cert_common_name, cert_issuer, cert_start_date, cert_end_date, cert_alternative_name) = parse_cert(row.tls_cert)`

Reply:
done